### PR TITLE
Jiadong - Replace Badges Section on Dashboard / badge count APIs

### PR DESCRIPTION
**Pull request by yunshan111**
# Description
image
image

## Related PRS (if any):
This frontend PR is related to the #[2146 backend PR.](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146/)
…

## Main changes explained:
Add badge count APIs
…

## How to test:
check into current branch and frontend [2146](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146/)
see frontend instruction

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HGNRest/assets/65985986/e0e974ab-ca4b-4fcd-b868-69f177cacc2d

### CODE REVIEW

**Review by hbodgal — APPROVED** [review]
image

Yes counter resets to 0. Works Fine.

**Comment by alla0810** [comment]
0. 'badge' tab not shown on development branch
1. checked out to Jiadong_Zhang_Replace_badge_Section branch.
2. errors while login as admin.
3. many errors resolved after waiting for around 10secs, but some of them still remain.
4. badge count updated successfully, and 'badge' tab appears after 'Weekly Summaries' tab.
5. But unlike above mentioned, red number indicator not shown
6. badge count reset after another badge assigned by admin

before

branch name

errors while login as admin

resolved after around 10sec

badge count not shown

**Review by aaryaneil — COMMENTED** [review]
The badge numbers didn't show up in the badge tab in red.

https://github.com/OneCommunityGlobal/HGNRest/assets/39597927/8cc7e8a0-ed9a-4166-aa32-749a587e77ce

**Review by aaryaneil — CHANGES_REQUESTED** [review]
The badge numbers didn't show up in the badge tab in red.

https://github.com/OneCommunityGlobal/HGNRest/assets/39597927/166901d9-0998-479f-a16a-c5608dcf5395

**Review by alla0810 — CHANGES_REQUESTED** [review]
errors while login as admin.
many errors resolved after waiting for around 10secs, but some of them still remain.
badge count updated successfully, and 'badge' tab appears after 'Weekly Summaries' tab.
But unlike above mentioned, red number indicator not shown

**Review by jennZhang93 — APPROVED** [review]
Left a review at the frontend PR with some UI suggestions: https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146#pullrequestreview-2010296515

**Review by SRamen1999 — COMMENTED** [review]
Left a review at the frontend PR https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146#pullrequestreview-2010423723

**Review by metaphor987 — CHANGES_REQUESTED** [review]
Hi Jiadong,

I've tested the changes using an admin account. The main functionalities are working well. I just encountered an issue here. If the admin assigns multiple badges to the user in one form(instead of clicking the "Confirm" to submit the "Assign Badge" form multiple times), the increased badge count in the user's dashboard will only increment by one. I've attached some screenshots below:

User account before receiving the badges:
Screenshot 2024-04-19 at 4 56 04 PM

Assigning badges process in the admin pages:
Screenshot 2024-04-19 at 4 57 43 PM

User account after badges are assigned:
Screenshot 2024-04-19 at 5 00 09 PM
Screenshot 2024-04-19 at 5 00 33 PM

Despite this issue, all other functionalities work very well. Thank you!

**Review by Tiantian-C — COMMENTED** [review]
I have tested this function, it works well!

https://github.com/OneCommunityGlobal/HGNRest/assets/122494725/1d55c3ab-af05-4b1e-8b3d-720b54115326

**Comment by yunshan111** [comment]
> adge count in the user's dashboard will o

Thank you, I am modifying it.

**Comment by yunshan111** [comment]
> 0. 'badge' tab not shown on development branch
> 1. checked out to Jiadong_Zhang_Replace_badge_Section branch.
> 2. errors while login as admin.
> 3. many errors resolved after waiting for around 10secs, but some of them still remain.
> 4. badge count updated successfully, and 'badge' tab appears after 'Weekly Summaries' tab.
> 5. But unlike above mentioned, red number indicator not shown
> 6. badge count reset after another badge assigned by admin
> 
> before
> 
> branch name
> 
> errors while login as admin
> 
> resolved after around 10sec
> 
> badge count not shown



> Hi Jiadong,
> 
> I've tested the changes using an admin account. The main functionalities are working well. I just encountered an issue here. If the admin assigns multiple badges to the user in one form(instead of clicking the "Confirm" to submit the "Assign Badge" form multiple times), the increased badge count in the user's dashboard will only increment by one. I've attached some screenshots below:
> 
> User account before receiving the badges: Screenshot 2024-04-19 at 4 56 04 PM
> 
> Assigning badges process in the admin pages: Screenshot 2024-04-19 at 4 57 43 PM
> 
> User account after badges are assigned: Screenshot 2024-04-19 at 5 00 09 PM Screenshot 2024-04-19 at 5 00 33 PM
> 
> Despite this issue, all other functionalities work very well. Thank you!

Thank you, I am fixing it

**Comment by yunshan111** [comment]
> errors while login as admin. many errors resolved after waiting for around 10secs, but some of them still remain. badge count updated successfully, and 'badge' tab appears after 'Weekly Summaries' tab. But unlike above mentioned, red number indicator not shown

Can you try it one more time? I cannot regenerate those bugs. backend: https://github.com/OneCommunityGlobal/HGNRest/tree/Jiadong_Zhang_Replace_badge_Section
frontend:https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/tree/Jiadong_Replace_Badges_Sectiong

**Comment by yunshan111** [comment]
> The badge numbers didn't show up in the badge tab in red.
> 
>  322160087-8cc7e8a0-ed9a-4166-aa32-749a587e77ce.mov

Can you show me the inspect so I can know what's happening, I cannot regenerate the bug
backend: https://github.com/OneCommunityGlobal/HGNRest/tree/Jiadong_Zhang_Replace_badge_Section
frontend:https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/tree/Jiadong_Replace_Badges_Sectiong

**Comment by mSharifHub** [comment]
Even when the put request fails due to a internal server error as shown in the browser console, the UI shows as updated successfully. 


Screenshot 2024-05-03 at 11 08 39 AM

File: src/controllers/badgeController.js [hunk 1]
```diff
@@ -262,12 +264,64 @@ const badgeController = function (Badge) {
...
+
+    UserProfile.findById(userId, (error, record) => {
+      // Check for errors or if user profile doesn't exist
+      if (error || record === null) {
+        res.sendStatus(404).send('Can not find the user to be assigned.');
+        return;
+      }
+      // Return badge count from user profile
+      res.status(200).send({ count: record.badgeCount });
+    });
```
Comment (mSharifHub) [inline src/controllers/badgeController.js:278]:
```suggestion
   const getBadgeCount = async function (req, res) {
    I recommend encapsulating in a try and catch
    
    try {
        const userId = mongoose.Types.ObjectId(req.params.userId);
     
        const userProfile = await UserProfile.findById(userId);

        if (!userProfile) {
            res.status(404).send('Cannot find the user to be assigned.');
            return;
        }
        res.status(200).send({ count: userProfile.badgeCount });
    } catch (error) {
        console.error("Error fetching user profile:", error);
        res.status(500).send('Server error occurred while fetching badge count.');
    }
};

```

**Review by mSharifHub — CHANGES_REQUESTED** [review]
There is an error which i took a screenshot in the comments section where we have an internal error and yet the user UI still displays that a badge was uploaded successfully.

**Review by Parth-tech — COMMENTED** [review]
Hi,
I have left a comment on the front end repo [link](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146#pullrequestreview-2039591500).

**Comment by KaushikMreddy** [comment]
Hi, 
I've left a review for this PR in the Frontend: https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146#pullrequestreview-2058763543

**Comment by peter6866** [comment]
Hi, I have left a review in the frontend PR #2146

**Review by Sandhya1236 — APPROVED** [review]
I checked out the current branch, ran npm install and the necessary commands to set up the PR locally. I logged in as an admin user and navigated to the dashboard -> Other Links -> Badge Management to assign a new badge to another user. However, the Badge Management page failed to load due to an error in the FilterBadges function within the BadgeDevelopmentTable.jsx
image

**Comment by dikshitakejriwal** [comment]
left a review in the frontend - https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146#pullrequestreview-2083956062

**Review by Infinite-Turkey — COMMENTED** [review]
Left a comment at Backend side [review](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2146#pullrequestreview-2087561039)

**Review by Vigneshwar-Muriki — APPROVED** [review]
Tested this PR and it works as expected.
Screenshot 2024-07-06 172717
Screenshot 2024-07-06 172631

**Pull request closed without merging**